R/main.R

```diff
@@ -15,37 +15,123 @@ check_drjacoby_loaded <- function() {
   x <- list(...)
   
-  # check input format
+  # check input format of arguments
   assert_gr(length(x), 0, message = "input cannot be empty")
-  if ((length(x) %% 4) != 0) {
-    stop("each parameter must have 4 inputs: {name, min, max, init}")
+  assert_in(names(x), c("name", "min", "max", "init", "block"))
+  use_init <- ("init" %in% names(x))
+  use_block <- ("block" %in% names(x))
+  n_cols <- 3 + use_init + use_block
+  if ((length(x) %% n_cols) != 0) {
+    stop("must have the same number of inputs per parameter")
+  }
+  n_param <- length(x) / n_cols
+  arg_names <-c("name", "min", "max")
+  if (use_init) {
+    arg_names <- c(arg_names, "init")
   }
-  n <- length(x) / 4
-  v <- 4*(0:(n-1))
-  mapply(function(x) {assert_single_string(x, message = "parameter names must be character strings")}, x[1 + v])
-  mapply(function(x) {assert_single_numeric(x, message = "min values must be single values")}, x[2 + v])
-  mapply(function(x) {assert_single_numeric(x, message = "max values must be single values")}, x[3 + v])
-  mapply(function(x) {assert_vector_numeric(x, message = "init values must be numeric")}, x[4 + v])
-  
-  # return dataframe
+  if (use_block) {
+    arg_names <- c(arg_names, "block")
+  }
+  assert_eq(names(x), rep(arg_names, n_param))
+  
+  # create params dataframe
+  v <- n_cols*(0:(n_param - 1))
   ret <- data.frame(name = unlist(x[1 + v]),
                     min = unlist(x[2 + v]),
                     max = unlist(x[3 + v]))
-  ret$init <- x[4 + v]
-  ret
+  if (use_init) {
+    ret$init <- x[4 + v]
+  }
+  if (use_block) {
+    ret$block <- x[5 + v]
+  }
+  
+  # run checks and standardise format
+  ret <- check_params(ret)
+  
+  return(ret)
+}
+
+#------------------------------------------------
+# Check that params dataframe formatted correctly, and return in standardised
+# format (init and block coerced to list)
+#' @noRd
+check_params <- function(x) {
+  
+  # check dataframe has correct elements
+  assert_dataframe(x)
+  assert_in(c("name", "min", "max"), names(x),
+            message = "df_params must contain the columns 'name', 'min', 'max'")
+  use_init <- ("init" %in% names(x))
+  use_block <- ("block" %in% names(x))
+  
+  # coerce init and block to list
+  if (use_init) {
+    if (!is.list(x$init)) {
+      x$init <- as.list(x$init)
+    }
+  }
+  if (use_block) {
+    if (!is.list(x$block)) {
+      x$block <- as.list(x$block)
+    }
+  }
+  
+  # check each row in turn
+  for (i in seq_len(nrow(x))) {
+    
+    # check format
+    assert_single_string(x$name[i], message = "parameter names must be character strings")
+    assert_single_numeric(x$min[i], message = "min values must be single values")
+    assert_single_numeric(x$max[i], message = "min values must be single values")
+    if (use_init) {
+      assert_vector_numeric(x$init[[i]], message = "init values must be numeric")
+    }
+    if (use_block) {
+      assert_vector_numeric(x$block[[i]], message = "block values must be numeric")

```

Comment:
I thought about this, and decided not really. In the C++ code we make a vector of length equal to the maximum value of $block. If a block ends up not being used it will just always contain 0, and contribute nothing to overall likelihood.